chronos/streams/asyncstream.nim

```diff
@@ -0,0 +1,908 @@
+    if resFut.read() != length:
+      raise newAsyncStreamIncompleteError()
+  else:
+    var item = WriteItem(kind: String)
+    if not isLiteral(sbytes):
+      shallowCopy(item.data3, sbytes)
+    else:
+      item.data3 = sbytes
+    item.size = length
+    item.future = newFuture[void]("async.stream.write(string)")
+    await wstream.queue.put(item)
+    yield item.future
+    if item.future.failed:
+      raise newAsyncStreamWriteError(item.future.error)
+
+proc finish*(wstream: AsyncStreamWriter) {.async.} =
+  ## Finish write stream ``wstream``.
+  if not wstream.running():
+    raise newAsyncStreamIncorrectError("Incorrect stream state")
+
+  if not isNil(wstream.wsource):
+    var item = WriteItem(kind: Pointer)
+    item.size = 0
+    item.future = newFuture[void]("async.stream.finish")
+    await wstream.queue.put(item)
+    yield item.future
+    if item.future.failed:
+      raise newAsyncStreamWriteError(item.future.error)
+
+proc join*(rw: AsyncStreamRW): Future[void] =
+  ## Get Future[void] which will be completed when stream become finished or
+  ## closed.
+  when rw is AsyncStreamReader:
+    var retFuture = newFuture[void]("async.stream.reader.join")
+  else:
+    var retFuture = newFuture[void]("async.stream.writer.join")
+  proc continuation(udata: pointer) = retFuture.complete()
+  if not rw.future.finished:
+    rw.future.addCallback(continuation)
+  else:
+    retFuture.complete()
+  return retFuture
+
+proc close*(rw: AsyncStreamRW) =
+  ## Close and frees resources of stream ``rw``.
+  ##
+  ## Note close() procedure is not completed immediately!
+  if rw.closed():
+    raise newAsyncStreamIncorrectError("Stream is already closed!")
+
+  proc continuation(udata: pointer) =
+    if not isNil(rw.udata):
+      GC_unref(cast[ref int](rw.udata))
+    rw.state = AsyncStreamState.Closed
+    rw.future.complete()
+    when rw is AsyncStreamReader:
+      untrackAsyncStreamReader(rw)
+    elif rw is AsyncStreamWriter:
+      untrackAsyncStreamWriter(rw)
+
+  when rw is AsyncStreamReader:
+    if isNil(rw.rsource):
+      callSoon(continuation)
+    else:
+      rw.exevent.fire()
+  elif rw is AsyncStreamWriter:
+    if isNil(rw.wsource):
+      callSoon(continuation)
+    else:
+      rw.exevent.fire()
+
+proc closeWait*(rw: AsyncStreamRW): Future[void] =
+  ## Close and frees resources of stream ``rw``.
+  rw.close()
+  result = rw.join()
+
+proc startReader(rstream: AsyncStreamReader) =
+  rstream.state = Running
+  if not isNil(rstream.readerLoop):
+    rstream.future = rstream.readerLoop(rstream)
+  else:
+    rstream.future = newFuture[void]("async.stream.empty.reader")
+
+proc startWriter(wstream: AsyncStreamWriter) =
+  wstream.state = Running
+  if not isNil(wstream.writerLoop):
+    wstream.future = wstream.writerLoop(wstream)
+  else:
+    wstream.future = newFuture[void]("async.stream.empty.writer")
+
+proc init*(child, wsource: AsyncStreamWriter, loop: StreamWriterLoop,
```

Comment:
I dont think we used `init` to allocate the instance, we always use `newXXX` to allocate new instance and `init` to initialize instance.

Reply:
https://github.com/status-im/nim-beacon-chain/blob/master/beacon_chain/beacon_node.nim#L92

Reply:
https://github.com/status-im/nim-beacon-chain/blob/master/beacon_chain/attestation_pool.nim#L8

this helps instantiation of types in a generic way, so that one can write type factories and the like..

Reply:
This examples are actually mistakes and do not follow common Nim's way. The common way is to use `initXXX` for objects, which are going to be allocated in stack and `newXXX` for objects which are going to be allocated in heap. I can give you many examples from Nim's library:

`initXXX`

* https://github.com/nim-lang/Nim/blob/devel/lib/pure/collections/heapqueue.nim#L58
* https://github.com/nim-lang/Nim/blob/devel/lib/pure/collections/tables.nim#L289
* https://github.com/nim-lang/Nim/blob/devel/lib/pure/collections/sets.nim#L123
* https://github.com/nim-lang/Nim/blob/devel/lib/pure/collections/sets.nim#L669

`newXXX`

* https://github.com/nim-lang/Nim/blob/devel/lib/pure/collections/tables.nim#L791

And this is nice example for usage of both `initXXX` and `newXXX`:

* https://github.com/nim-lang/Nim/blob/e1515b53d1992aa8443a3317759cf5bab7fa9139/lib/pure/collections/lists.nim

Reply:
In the rest of the Status codebases, we tend to follow this proposal:
https://github.com/nim-lang/RFCs/issues/48

The goal is to make the constructors easier to use in generic context.

Reply:
@zah sorry, but its first i see this RFC and i'm curious who else knows about this?
https://github.com/search?q=org%3Astatus-im+proc+init&type=Code

Reply:
:) Eugene, is your memory failing you?

I remember telling you about this convention in a Gitter chat once and you are using it in nim-libp2p for example:

https://github.com/status-im/nim-libp2p/blob/eed7ce92c1157b30846d8bb5fa199f21c9d1b9aa/libp2p/crypto/minasn1.nim#L629

Reply:
I think you missed point of this discussion @zah. The point here is that `init` is used to initalize instance, but not allocate object.

Reply:
I see. Mamy has suggested that we can use the same convention for allocating procs, but with `new` instead of `init`:

```nim
proc new(T: type Foo, x, y: Bar): ref Foo =
  ...
```

Reply:
I can't make up my mind on that one.. on the one hand, it's kind of convenient to always use the same syntax for construction ('init'), but on the other, there are significant ways to shoot yourself in the foot if you forget that a type is `ref object`, specially when copying them around.. perhaps `new(T: type..)` is safer for this reason and we should differentiate the two.. too bad there's still no way to enforce that it's actually called.